src/UndisclosedPassword.php

```diff
@@ -0,0 +1,177 @@
+<?php
+
+
+namespace Zend\Validator;
+
+use Psr\Http\Client\ClientExceptionInterface;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+
+final class UndisclosedPassword extends AbstractValidator
+{
+    const HIBP_API_URI = 'https://api.pwnedpasswords.com';
+    const HIBP_API_TIMEOUT = 300;
+    const HIBP_CLIENT_UA = 'zend-validator';
+    const HIBP_CLIENT_ACCEPT = 'application/vnd.haveibeenpwned.v2+json';
+    const HIBP_RANGE_LENGTH = 5;
+    const HIBP_RANGE_BASE = 0;
+    const HIBP_COUNT_BASE = 0;
+    const SHA1_LENGTH = 40;
+
+    const PASSWORD_BREACHED = 'passwordBreached';
+    const WRONG_INPUT = 'wrongInput';
+    const CONNECTION_FAILURE = 'connFail';
+    const UNKNOWN_ERROR = 'unknownError';
+
+    protected $messageTemplates = [
+        self::PASSWORD_BREACHED => 'The provided password was used by others',
+        self::WRONG_INPUT => 'The provided password failed to be correctly hashed, please verify your input',
+        self::CONNECTION_FAILURE => 'Unable to reach HIBP service, please try again later',
+        self::UNKNOWN_ERROR => 'Something happened beyond our control, error reporting should give more details',
+    ];
+
+    /**
+     * @var ClientInterface
+     */
+    private $httpClient;
+
+    /**
+     * @var RequestFactoryInterface
+     */
+    private $httpRequest;
+
+    /**
+     * @var ResponseFactoryInterface
+     */
+    private $httpResponse;
+
+    /**
+     * @var int
+     */
+    private $count = 0;
+
+    /**
+     * PasswordBreach constructor.
+     *
+     * @param ClientInterface $httpClient
+     * @param RequestFactoryInterface $httpRequest
+     * @param ResponseFactoryInterface $httpResponse
+     */
+    public function __construct(
+        ClientInterface $httpClient,
+        RequestFactoryInterface $httpRequest,
+        ResponseFactoryInterface $httpResponse
+    ) {
+        $this->httpClient = $httpClient;
+        $this->httpRequest = $httpRequest;
+        $this->httpResponse = $httpResponse;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function isValid($value)
+    {
+        if (! is_string($value)) {
+            $this->error(self::WRONG_INPUT);
+            return false;
+        }
+        try {
+            $isPwnd = $this->isPwnedPassword($value);
+        } catch (Exception\RuntimeException $runtimeException) {
+            $this->error(self::CONNECTION_FAILURE);
+            return false;
+        } catch (\Exception $exception) {
+            $this->error(self::UNKNOWN_ERROR);
+            return false;
+        }
+        if ($isPwnd) {
+            $this->error(self::PASSWORD_BREACHED);
+            return false;
+        }
+        return true;
+    }
+
+    private function isPwnedPassword($password)
```

Comment:
Noticed while looking at the lack of type declarations: new code should target `develop` (target branch needs changing in the PR)

Reply:
Moved base to `develop` but even there the composer refers to php 5.6 so until your proposal is agreed upon, let's leave out the type declarations for now. It's a jiffy to put them in, but I'm already struggling with a failed pass on PHP 5.6 because of PSR18 and I'd like to tackle one issue at the time. Let's solve zendframework/maintainers#59 first before adding type declarations.